### ISSUE TITLE
Support gradient clipping

### ISSUE THREAD

**Issue by MichaelKonobeev** [facebookresearch/higher#18]
When using differentiable optimizer we call `diffopt.step(loss)` which updates the parameters of a functional model and returns them, therefore it seems not possible to do gradient clipping. Would it be a good idea to add a function-type argument to the step method of differentiable optimizers which takes as input either [`all_grads`](https://github.com/facebookresearch/higher/blob/master/higher/optim.py#L208) or [`grouped_grads`](https://github.com/facebookresearch/higher/blob/master/higher/optim.py#L223) and should returned modified gradients or is there a better way of doing this? I could work on a PR implementing this given some guidance.

**Comment by egrefen** [comment]
Hello. I've returned from leave and allocated some time to look into this issue over the next two weeks. Hopefully we'll make some progress and report back, or come back to you with questions.

**Comment by egrefen** [comment]
@MichaelKonobeev could you please take a look at #21 and see if it fits your needs? Feedback/improvement suggestions welcome.

Please note that if you literally clip gradients, it will not be possible to take higher order gradients (backprop through backprop) as they will not exist. A differentiable continuous relaxation of the clipping operation will be required.

**Comment by MichaelKonobeev** [comment]
The problem with higher order gradients not existing occurs in the case of [clipping by value](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L50), but it does not seem to me that there is an issue in the case of [clipping by norm](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L6). Assuming that's correct what are your thoughts on adding functions which could help with clipping in a non in-place way, as I've done [here](https://github.com/MichaelKonobeev/higher/blob/master/higher/clip.py)? I could add this with a PR if that's ok with you.

**Comment by AntoineHX** [comment]
> The problem with higher order gradients not existing occurs in the case of [clipping by value](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L50), but it does not seem to me that there is an issue in the case of [clipping by norm](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L6). Assuming that's correct what are your thoughts on adding functions which could help with clipping in a non in-place way, as I've done [here](https://github.com/MichaelKonobeev/higher/blob/master/higher/clip.py)? I could add this with a PR if that's ok with you.

@MichaelKonobeev Wouldn't be more useful to support the gradient clipping function from PyTorch (which takes parameters rather than gradients as inputs) than adding your function ? I believe your function goal is to mimic the behavior of the [clipping by norm](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L6), right ?

**Comment by egrefen** [comment]
@MichaelKonobeev Your utility functions would be most welcome contributions in `higher.utils`, if you would like to submit a separate PR. It would be helpful if minimal tests for these functions were added to `tests/utils.py`.

For now, I am closing this issue as I am assuming it is addressed by #21. Please re-open if it does not address the issue, or flag new problems in a separate issue.
